### PR TITLE
Add slot to va-text-input

### PR DESCRIPTION
**Pull request by andresriveratoro**
## Chromatic

https://35734-va-text-input-label--60f9b557105290003b387cd5.chromatic.com

## Description
Closes https://github.com/department-of-veterans-affairs/va.gov-team/issues/35734

## Testing done
E2E and Storybook

## Screenshots
Screen Shot 2022-03-31 at 18 08 20

## Acceptance criteria
- [x] Add ability to pass in HTMLElements to be used for label

## Definition of done
- [ ] Documentation has been updated, if applicable
- [x] A link has been provided to the originating GitHub issue (or connected to it via ZenHub)

### CODE REVIEW

File: packages/storybook/stories/va-text-input.stories.jsx
Comment (andresriveratoro) [inline packages/storybook/stories/va-text-input.stories.jsx:27]: These were updated to `undefined` to prevent them from showing up in Storybook's `Show Code` as `error={null}`.

File: packages/web-components/src/components/va-text-input/va-text-input.tsx
Comment (andresriveratoro) [inline packages/web-components/src/components/va-text-input/va-text-input.tsx:51]: The type was updated to have better documentation in Storybook.

File: packages/web-components/src/components/va-text-input/va-text-input.tsx
Comment (andresriveratoro) [inline packages/web-components/src/components/va-text-input/va-text-input.tsx:57]: See comment for `inputmode`.

File: packages/web-components/src/components/va-text-input/va-text-input.tsx
Comment (andresriveratoro) [inline packages/web-components/src/components/va-text-input/va-text-input.tsx:167]: This was done so engineers have the option to use a label and insert content between the label and input field. Engineers can also bypass the label property to have finer control over what is rendered above the input field.

**Comment by pdavies88** [comment]
> I can remove the Experimental story if we don't want that example here.

I mean I think we should have a story I don't love the experimental name was thinking more along the lines of "HTML within Text Input" just like what it is. I mean at the core of this ticket I would much rather have developers put this HTML before the component than inside it but thats not part of the request lol.

**Review by pdavies88 — APPROVED** [review]

**Comment by k80bowman** [comment]
Just to make sure I understand what you've done here, the implementation on this is changing. The label prop now only accepts text, but engineers can also skip use of the label prop and insert an HTML element instead. Or they could have a label and an HTML element. Is this correct?

If so, how will this affect implementations of this component that are already in place?

**Comment by pdavies88** [comment]
Ah gotcha yeah maybe we should put a link to that in the notes at the top or change the name. Cause without that context I was kinda like this name is odd cause we are recommending this in our official storybook so it makes it like not experimental at that point.

**Comment by pdavies88** [comment]
@k80bowman @andresriveratoro side note if we implement this should we remove this page from the docs: https://design.va.gov/experimental-design/hint-text#within-an-additional-info-component now that implementation would be part of the official code base?

**Comment by andresriveratoro** [comment]
I only included the story because without it engineers won't have an example using the slot.

Here's a `TextInput` label used in vets-website:
```jsx
  <>
    <strong>Name of issue</strong>
    <span className="form-required-span">(*Required)</span>
    <p className="vads-u-font-weight--normal label-description">
      You can only add an issue that you’ve already received a VA decision
      notice for.
    </p>
  </>
```

Using `va-text-input`:
```jsx
<va-text-input label="Name of issue" {...otherProps}>
  <p className="vads-u-font-weight--normal label-description">
    You can only add an issue that you’ve already received a VA decision notice for.
  </p>
</va-text-input>
```

Should I update the story to this and rename it?

**Comment by pdavies88** [comment]
IMO yes we should rename and provide a concrete example that a developer on another team could run with and allow them to move away from the react component cause that's the ultimate goal.

**Comment by k80bowman** [comment]
I'm a little nervous about implementing an experimental design in the design system without consulting the DSC. Let's hold off on merging this until I can talk with Carol and make sure this is OK. We may need to bring it up with the DSC before merging.

I also think we should rename the story and provide some more guidance on when and how this feature should be used.

**Comment by andresriveratoro** [comment]
I just realized that last example is also under our Experimental Design, but it is being used in `vets-website`.

File: packages/web-components/src/components/va-text-input/va-text-input.tsx
Comment (bkjohnson) [reply packages/web-components/src/components/va-text-input/va-text-input.tsx:167]: This seems fine, holding off on approval for an a11y :+1:

File: packages/web-components/src/components/va-text-input/test/va-text-input.e2e.ts
Comment (bkjohnson) [inline packages/web-components/src/components/va-text-input/test/va-text-input.e2e.ts:16]: This is in the `renders without a label` test - should this be happening?

Comment (andresriveratoro) [reply packages/web-components/src/components/va-text-input/test/va-text-input.e2e.ts]: I've updated this test to use a `label` prop and added a separate test when the slot is used as a label.

Comment (andresriveratoro) [reply packages/web-components/src/components/va-text-input/test/va-text-input.e2e.ts]: I updated the test because a developer would be providing either a `label` prop string or children to be used inside the label tag. The original render test was updated to cover the case where a label prop is used.

**Review by bkjohnson — COMMENTED** [review]
We should probably have @CourtneyChambers77 take a look at this.

**Comment by andresriveratoro** [comment]
I updated the default story to make it easier for @CourtneyChambers77 to do an accessibility test using slot content in the label. I'll revert this before merging this PR.

**Comment by andresriveratoro** [comment]
The story changes have been reverted. Having the slot inside the label is not a problem as long as a few guidelines are followed. According to the [spec](https://html.spec.whatwg.org/multipage/forms.html#the-label-element), acceptable children are "[Phrasing content](https://html.spec.whatwg.org/multipage/dom.html#phrasing-content-2), but with no descendant [labelable elements](https://html.spec.whatwg.org/multipage/forms.html#category-label) unless it is the element's [labeled control](https://html.spec.whatwg.org/multipage/forms.html#labeled-control), and no descendant [label](https://html.spec.whatwg.org/multipage/forms.html#the-label-element) elements." 

Using the experimental design example in this way would be going against the spec because `va-additional-info` is not a form control and it contains a `div`.

The goal of this issue was to give developers more control over what is displayed and how it's displayed before the input field. If we move the `slot` outside of the label **element** and engineers don't provide a `label` **prop**, the responsibility of adding the label element and any other markup falls to them but they would have all the flexibility they would need. If we leave the slot inside the label, engineers could potentially insert elements that go against the spec. I prefer the first option, but would love to know your thoughts @bkjohnson @pdavies88 @k80bowman @mendacitees.

**Comment by pdavies88** [comment]
@andresriveratoro I would agree with you to err on the side of caution and rather have them do the html prior instead of going against the spec. However I do think that would invalidate the ticket and the experimental design so there's that to think about as well. Maybe we want to add it to the Monday meeting

**Comment by andresriveratoro** [comment]
It looks like we have a critical accessibility error when passing in `<label htmlFor="inputField">` from the light DOM - the input is never linked to this label.  

I modified the code to this:
```jsx
{this.label && (
  <label htmlFor="inputField">
    <Fragment>
      {this.label}{' '}
      {this.required && <span class="required">(*Required)</span>}
    </Fragment>
  </label>
)}
<slot></slot>
```

The experimental design would work with this change when a label is provided. 
Screen Shot 2022-03-31 at 10 46 45

File: packages/web-components/src/components/va-text-input/va-text-input.tsx
Comment (andresriveratoro) [reply packages/web-components/src/components/va-text-input/va-text-input.tsx:161]: Removed

**Comment by andresriveratoro** [comment]
@pdavies88 I've updated the story name to 'With Hint Text'.

**Review by bkjohnson — APPROVED** [review]
:shipit:

**Pull request merged**
